packages/hardhat-create2-proxy/src/deployCreate2Factory.ts

```diff
@@ -0,0 +1,206 @@
+import {
+  getCreateAddress,
+  keccak256,
+  Signer,
+  toUtf8Bytes,
+  Wallet,
+} from 'ethers';
+
+import create2Artifact from '../build/create2-factory/combined.json';
+import hre, { chainweb, ethers } from 'hardhat';
+import { getNetworkStem } from '@kadena/hardhat-chainweb';
+
+const networkStem = getNetworkStem(hre.config.defaultChainweb);
+
+function isContractDeployed(address: string): Promise<boolean> {
+  return ethers.provider.getCode(address).then((code) => code !== '0x');
+}
+
+export const create2Artifacts =
+  create2Artifact.contracts['contracts/Create2Factory.sol:Create2Factory'];
+
+export const getCreate2FactoryAddress = async (
+  signer: Signer,
+  version: number | bigint = BigInt(1),
+) => {
+  const secondaryKey = await deriveSecondaryKey(signer, version);
+  const factoryAddress = getCreateAddress({
+    from: secondaryKey.publicKey,
+    nonce: 0,
+  });
+
+  return factoryAddress;
+};
+
+export async function deriveSecondaryKey(
+  signer: Signer,
+  version: number | bigint = BigInt(1),
+) {
+  const message = `create deployer key for create2 factory version: ${version}`;
+  const signature = await signer.signMessage(message);
+
+  // Combine signature and label to get deterministic entropy
+  const hash = keccak256(toUtf8Bytes(signature));
+
+  // Use first 32 bytes (64 hex chars + '0x') as the private key
+  const derivedPrivateKey = '0x' + hash.slice(2, 66);
+  const wallet: Wallet = new Wallet(derivedPrivateKey, ethers.provider);
+  return {
+    publicKey: await wallet.getAddress(),
+    privateKey: derivedPrivateKey,
+  };
+}
+
+async function fundDeployer(
+  sender: Signer,
+  receiver: Signer,
+  amountStr: string,
+) {
+  const amount = ethers.parseEther(amountStr); // 1 ether
+
+  const receiverAddress = await receiver.getAddress();
+
+  const tx = await sender.sendTransaction({
+    to: receiverAddress,
+    value: amount,
+  });
+  await tx.wait();
+
+  const receiverBalance = await ethers.provider.getBalance(receiverAddress);
+  if (receiverBalance !== amount) {
+    throw new Error(
+      `funding deployer failed. receiver balance: ${receiverBalance} is not equal to funding amount: ${amount}`,
+    );
+  }
+}
+
+export const deployCreate2Factory = async (props?: {
+  signer?: string;
+  version?: number;
+  fundingDeployerWith?: string;
+}) => {
+  const { signer, version = 1, fundingDeployerWith = '1.0' } = props ?? {};
```

Comment:
See results from running the benchmark-create2Factory-gas.ts

=== RECOMMENDED VALUE ===
Recommended 'fundingDeployerWith' value: '0.00523377' ETH
This includes:
- Average gas used: 174459 gas units
- At a gas price of 20 gwei
- With a 50% safety buffer for network congestion

You can runn it against localhost to see all the output from the 5 runs.
So 0.005 (a bit higher than recommended) should be sufficient.
We can run the script against devnet when it is back up and later against testnet to be sure.
But 1.0 is definitely too much

Reply:
Modified